system/lib/wasmfs/proxied_file_backend.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the Proxied File and Proxied Backend of the new file
+// system. Current Status: Work in Progress. See
+// https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "backend.h"
+#include "file.h"
+#include "thread_utils.h"
+#include "wasmfs.h"
+
+namespace wasmfs {
+
+// This class represents a file that is resident on a background pthread. All
+// file operations are proxied to this dedicated thread.
+class ProxiedFile : public DataFile {
+
+  emscripten::SyncToAsync& proxy;
+  std::shared_ptr<DataFile> backendFile;
+
+  // Read and write operations are forwarded to the file residing on the
+  // dedicated thread.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+    __wasi_errno_t result;
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      result = backendFile->locked().write(buf, len, offset);
+      (*resume)();
+    });
+    return result;
+  }
+
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    __wasi_errno_t result;
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      result = backendFile->locked().read(buf, len, offset);
+      (*resume)();
+    });
+    return result;
+  }
+
+  // Querying the size of the Proxied File returns the size of the underlying
+  // file on the dedicated thread.
+  size_t getSize() override {
+    size_t result;
+    proxy.invoke([&](emscripten::SyncToAsync::Callback resume) {
+      result = backendFile->locked().getSize();
+      (*resume)();
+    });
+    return result;
+  }
+
+public:
+  // A file with the chosen destination backend is created on the Proxied File's
+  // dedicated thread.
+  ProxiedFile(mode_t mode,
+              backend_t srcBackend,
+              backend_t destBackend,
+              emscripten::SyncToAsync& proxy)
```

Comment:
Aside from the names of src/destBackend, I'm not sure what these are - a comment would be good.

On that topic, why does this receive a backend at all? It seems like it could receive the SyncToAsync and nothing else. That would leave this class as doing the minimal amount of things: it literally can get just a file + a SyncToAsync and to forward all ops using that. How does that sound?

Reply:
Hmm, so we would create the file itself in `invoke` and then pass that file to the constructor? I think that would be nicer.